Add more XElement extensions and Xml utilities

regnrand opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/libpalaso/519)
<!-- Reviewable:end -->

jasonleenaylor commented on SIL.Core/Xml/XmlUtils.cs:
```diff
@@ -762,5 +1191,159 @@ private static bool IsLegalXmlChar(int codePoint)
+	/// <summary>
+	/// Interface for operations we can apply to attributes.
+	/// </summary>
+	public interface IAttributeVisitor
```
If ReplaceSubstringInAttr is the only IAttributeVisitor then this seems overengineered. VisitAttributes could be replaced by a ReplaceSubstringInAttributes method.

regnrand commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[SIL.Core/Xml/XmlUtils.cs, line 1325 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkCuXn2ay7WQZHoAWSz-r1-1325:-KkE87tI9JKzAkmZaNdW:b-b8s6j7) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L1325)):*
<details><summary><i>Previously, jasonleenaylor (Jason Naylor) wrote…</i></summary><blockquote>

If ReplaceSubstringInAttr is the only IAttributeVisitor then this seems overengineered. VisitAttributes could be replaced by a ReplaceSubstringInAttributes method. 
</blockquote></details>

FLEx has four other implementations of the IAttributeVisitor interface, so it looks to be loved beyond this file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/519)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
I realize most/all of this comes over direct from FLEx, and most of my comments could be considered nitpicking/preference, but they are based on either points of confusion for me as I read the code or my understanding of our conventions or my understanding of best practices.

Up to you which you think are warranted.

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 186 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGEzsYUdDWjW4HThBd:-KkGEzsZQlLCLg-EXVFK:b-bqb0uw) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L186)):*
> ```Smalltalk
> 		/// </summary>
> 		[Test]
> 		public void CloneIsTheSameAsSource()
> ```

This could be a little cleaner if each of these calls was simple a [TestCase] on DoCloneAndCheckIt (which would then warrant a rename). Then each would run and report independently.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 203 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGFSvLbxlvtdfqXsuq:-KkGFSvLbxlvtdfqXsur:b9cq0p0) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L203)):*
> Assert.AreEqual(clone.ToString(), sourceElement.ToString());
 
Tehnically, this should be Assert.AreEqual(sourceElement.ToString(), clone.ToString());

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 204 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGFgiwmSx0qRrfrAMQ:-KkGFgiwmSx0qRrfrAMR:b-fuqs22) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L204)):*
> StringAssert.Contains(testData, clone.ToString());

Doesn't hurt anything, but this check seems extraneous.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 211 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGGM4aBHB_iuu0ly5P:-KkGGM4aBHB_iuu0ly5Q:bn2xz4i) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L211)):*
> ```Smalltalk
> 		/// </summary>
> 		[Test]
> 		public void OuterXmlIsSameAsToStringText()
> ```

By convention, we usually write test names as GetOuterXml_xxx so it is clear what method we are testing. It would help me understand especially in a case like this. Otherwise, it is unclear if we are testing GetOuterXml or ToString.
Personally, I would advocate for this change to all these tests.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 215 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGHAfuKK7_BmszxTtl:-KkGHAfvHvfEia1CiD_G:bt0bqln) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L215)):*
> ```Smalltalk
> 			const string outerXmlData = "<stuff attr='myAttrValue' ><!-- Some comment --><child attr='myChildAttrValue' /></stuff>";
> 			const string testData = "<!-- Some comment -->";
> 			StringAssert.Contains(testData, outerXmlData);
> ```

Is this just a sanity check? Seems extraneous.
Same for tests below.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 228 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGHyPl8X2AxsRB_Jpc:-KkGHyPl8X2AxsRB_Jpd:bzcqw3t) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L228)):*
> testData

I think it would help my brain parse these tests if this was "expected" or "expectedResult".

Also, it would be more clear what is being tested here if these were two tests. Then the one with the comment could include language describing the expectation that comments aren't included in the method name. Same below.

---

*[SIL.Core.Tests/Xml/XmlUtilsTests.cs, line 178 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGKvvtDYjmovOPoLu-:-KkGKvvtDYjmovOPoLu0:bjmobwo) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XmlUtilsTests.cs#L178)):*
> "Null returns false"

I think this would confuse me if I saw it in relation to a failing test. How about, "Null should return false"?
Same for the rest.

This would be a perfect candidate for [TestCase]s

---

*[SIL.Core/Xml/XmlUtils.cs, line 17 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGTjHRr-P_dGz7t_zv:-KkGTjHRr-P_dGz7t_zw:bftmu2k) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L17)):*
> ```Smalltalk
> 	/// Summary description for XmlUtils.
> 	/// </summary>
> 	public static class XmlUtils
> ```

The new methods in here could use tests.

---

*[SIL.Core/Xml/XmlUtils.cs, line 201 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGMCtYBwQ6PEck6V9y:-KkGMCtYBwQ6PEck6V9z:batz1m1) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L201)):*
> ```Smalltalk
> 		/// <param name="attrName"></param>
> 		/// <returns></returns>
> 		public static uint[] GetMandatoryUIntegerListAttributeValue(XmlNode node, string attrName)
> ```

Should this be marked CLS-compliant false?
Same for MakeListValue

---

*[SIL.Core/Xml/XmlUtils.cs, line 256 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGMYtX9ZYEXKkVHFLU:-KkGMYtX9ZYEXKkVHFLV:b-jlgr20) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L256)):*
> MakeListValue

This name seems awkward. How about something like ListToString or GetStringFromList or MakeStringFromList?

---

*[SIL.Core/Xml/XmlUtils.cs, line 481 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGORDxSGl2MQTIFLCh:-KkGORDxSGl2MQTIFLCi:b-at1k5) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L481)):*
> ```Smalltalk
> 		/// <param name="name">The XElement name to find.</param>
> 		/// <returns></returns>
> 		public static XElement FindElement(XElement element, string name)
> ```

Perhaps I'm missing it, but this method seems identical to the one above.

---

*[SIL.Core/Xml/XmlUtils.cs, line 499 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGOrSUCNDsP-9tWnjT:-KkGOrSUCNDsP-9tWnjU:b-bc096l) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L499)):*
> Return -1 if not found.

Belongs in the &lt;returns&gt;&lt;/returns&gt; below

---

*[SIL.Core/Xml/XmlUtils.cs, line 589 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGPSDFG7D01cPUlUDz:-KkGPSDFG7D01cPUlUE-:b-y7x2rf) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L589)):*
> ```Smalltalk
> 		/// <param name="attrName"></param>
> 		/// <param name="attrVal"></param>
> 		public static void AppendAttribute(XElement parent, string attrName, string attrVal)
> ```

Is an attribute's element considered its parent? This confused me. 
I would replace "parent" with "element" in code and comment.
Also, "AppendAttribute" seems a little confusing. How about AddOrUpdateAttribute or AddOrSetAttribute or SetAttribute?

I now see you have SetAttribute below. Why do we need both? If this one is only supposed to add, it should be AddAttribute and should not update an existing one.

---

*[SIL.Core/Xml/XmlUtils.cs, line 868 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGRH9MUNHU4FILn--l:-KkGRH9MUNHU4FILn--m:bgqdr2w) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L868)):*
> ```Smalltalk
> 		/// <summary>
> 		/// Convert a possibly multiparagraph string to a form that is safe to store both in an XML file. Also deals
> 		/// with some characters that are not safe to put in an FDO field, which is more to the point since that's
> ```

Looks like this comment could use some cleanup now that it doesn't reside in Fieldworks.

---

*[SIL.Core/Xml/XmlUtils.cs, line 1209 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGS--slmJkOevmFxZx:-KkGS--slmJkOevmFxZy:b-td77ec) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L1209)):*
> ```Smalltalk
> 				{
> 					if (visitor.Visit(new XAttribute(xa.Name, xa.Value)))
> 						fSuccessfulVisit = true;
> ```

Why not return true here? And in the loop below?
Then return false at the end.

Same in next method

---

*[SIL.Core/Xml/XmlUtils.cs, line 1258 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGU35mmDEinu3l5jAv:-KkGU35n_rqvXPOtUidn:b-1bk4sy) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L1258)):*
> ```Smalltalk
> 		/// <param name="objToSerialize"></param>
> 		/// <returns>empty string if couldn't serialize object</returns>
> 		public static string SerializeObjectToXmlString(object objToSerialize)
> ```

This should probably be in XMLSerializationHelper.
Same with next method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-:-KkGUIEvynkIQqpL7Ov9:bosw2dn)*
<!-- Sent from Reviewable.io -->

regnrand commented:
Review status: 5 of 8 files reviewed at latest revision, 17 unresolved discussions.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 186 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGEzsYUdDWjW4HThBd:-KkGZWOL2W23IYoldQFi:b-a3806e) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L186)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

This could be a little cleaner if each of these calls was simple a [TestCase] on DoCloneAndCheckIt (which would then warrant a rename). Then each would run and report independently.
</blockquote></details>

Done. I assume it was written eons ago, before [TestCase] was invented. Also, I can't say that I'm current on the newer stuff in the test framework.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 203 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGFSvLbxlvtdfqXsuq:-KkGZjW8EMs69Rf5bBtR:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L203)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> Assert.AreEqual(clone.ToString(), sourceElement.ToString());
 
Tehnically, this should be Assert.AreEqual(sourceElement.ToString(), clone.ToString());
</blockquote></details>

Done.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 204 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGFgiwmSx0qRrfrAMQ:-KkGZpoPPPn-KylbDsbZ:bss4a2b) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L204)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> StringAssert.Contains(testData, clone.ToString());

Doesn't hurt anything, but this check seems extraneous.
</blockquote></details>

Removed.

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 211 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGGM4aBHB_iuu0ly5P:-KkG_VzvZX1nAOoz0yRb:b-u9jilt) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L211)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

By convention, we usually write test names as GetOuterXml_xxx so it is clear what method we are testing. It would help me understand especially in a case like this. Otherwise, it is unclear if we are testing GetOuterXml or ToString.
Personally, I would advocate for this change to all these tests.
</blockquote></details>

Done (x3).

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 215 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGHAfuKK7_BmszxTtl:-KkGaIx7XNURftpz9MFu:b-unvzb1) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L215)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is this just a sanity check? Seems extraneous.
Same for tests below.
</blockquote></details>

Removed (x3).

---

*[SIL.Core.Tests/Xml/XElementExtensionTests.cs, line 228 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGHyPl8X2AxsRB_Jpc:-KkGcgzy343I6UugtqiH:blykkj0) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XElementExtensionTests.cs#L228)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> testData

I think it would help my brain parse these tests if this was "expected" or "expectedResult".

Also, it would be more clear what is being tested here if these were two tests. Then the one with the comment could include language describing the expectation that comments aren't included in the method name. Same below.
</blockquote></details>

1. Used "expectedResult"
2. Split test into two (with and w/o comment (x2).

---

*[SIL.Core.Tests/Xml/XmlUtilsTests.cs, line 178 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGKvvtDYjmovOPoLu-:-KkGgKB5ejuHi2zbKkeA:b-kmge3h) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core.Tests/Xml/XmlUtilsTests.cs#L178)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> "Null returns false"

I think this would confuse me if I saw it in relation to a failing test. How about, "Null should return false"?
Same for the rest.

This would be a perfect candidate for [TestCase]s
</blockquote></details>

1. All error messages changed
2. Using [TestCase].

---

*[SIL.Core/Xml/XmlUtils.cs, line 17 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGTjHRr-P_dGz7t_zv:-KkIUvtRh6heyKSHn-EQ:b2yoovj) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L17)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

The new methods in here could use tests.
</blockquote></details>

I added a lot of new tests. Can't say if I got them all.

---

*[SIL.Core/Xml/XmlUtils.cs, line 201 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGMCtYBwQ6PEck6V9y:-KkGp9J1mIH4o8OmoBjE:b-u9jjci) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L201)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Should this be marked CLS-compliant false?
Same for MakeListValue
</blockquote></details>

Done (x2).

---

*[SIL.Core/Xml/XmlUtils.cs, line 256 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGMYtX9ZYEXKkVHFLU:-KkGpaNBMbrXYB0N4TyW:b-u9jjci) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L256)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> MakeListValue

This name seems awkward. How about something like ListToString or GetStringFromList or MakeStringFromList?
</blockquote></details>

Done (x2).

---

*[SIL.Core/Xml/XmlUtils.cs, line 481 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGORDxSGl2MQTIFLCh:-KkGq-EDrGBFKc9DRyto:b-cgfrh5) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L481)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Perhaps I'm missing it, but this method seems identical to the one above.
</blockquote></details>

Removed the XElement method called FindNode.

---

*[SIL.Core/Xml/XmlUtils.cs, line 499 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGOrSUCNDsP-9tWnjT:-KkGqH9Te2eFiNJnI5uN:b-zckbes) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L499)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> Return -1 if not found.

Belongs in the &lt;returns&gt;&lt;/returns&gt; below
</blockquote></details>

Moved to <returns></returns>.

---

*[SIL.Core/Xml/XmlUtils.cs, line 589 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGPSDFG7D01cPUlUDz:-KkGr76AqPdPMdpwt1rj:b8h9uxh) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L589)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is an attribute's element considered its parent? This confused me. 
I would replace "parent" with "element" in code and comment.
Also, "AppendAttribute" seems a little confusing. How about AddOrUpdateAttribute or AddOrSetAttribute or SetAttribute?

I now see you have SetAttribute below. Why do we need both? If this one is only supposed to add, it should be AddAttribute and should not update an existing one.
</blockquote></details>

1. Renamed (x2) (also for XmlNode overload).
2. Kept overloads of ''SetAttribute" and removed the overloads of AppendAttribute.

---

*[SIL.Core/Xml/XmlUtils.cs, line 868 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGRH9MUNHU4FILn--l:-KkHSK2VoKy_C9GJnkba:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L868)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Looks like this comment could use some cleanup now that it doesn't reside in Fieldworks.
</blockquote></details>

Done.

---

*[SIL.Core/Xml/XmlUtils.cs, line 1209 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGS--slmJkOevmFxZx:-KkHT2iWzxwsu7K-Kw1s:b-1k3vr3) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L1209)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Why not return true here? And in the loop below?
Then return false at the end.

Same in next method
</blockquote></details>

Dunno. I just moved the code. I do know that FLEX has several more implementations of the IAttributeVisitor interface, and I'm reluctant to change so much behavior, without a better understanding of those return values.

---

*[SIL.Core/Xml/XmlUtils.cs, line 1258 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGU35mmDEinu3l5jAv:-KkHWegYd6KN8aPOSAyi:b-jg3cne) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L1258)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

This should probably be in XMLSerializationHelper.
Same with next method.
</blockquote></details>

I just removed "SerializeObjectToXmlString" & "DeserializeXmlString", since it looks like there are numerous methods in XMLSerializationHelper to do that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/519)*
<!-- Sent from Reviewable.io -->

jasonleenaylor commented:
I'd like to see the variable in the VisitAttribute match modern name conventions, otherwise LGTM.

---

Reviewed 3 of 10 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-:-KkL8alzNsyZuaqIX7h2:b-3ysffz)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one nit thing remaining (an empty method in a test class).
Thanks for doing this!

---

Reviewed 1 of 10 files at r1, 2 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[SIL.Core.Tests/Xml/XmlUtilsTests.cs, line 368 at r2](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkLHuMumUw5qInW7-cj:-KkLHuMvj2YzusZUTjjs:b-5ncrdr) ([raw file](https://github.com/sillsdev/libpalaso/blob/6626cc904af08d674ef27f9322048efca49d50d1/SIL.Core.Tests/Xml/XmlUtilsTests.cs#L368)):*
> ```Smalltalk
> 		}
> 
> 		public void DecodeXmlTest()
> ```

Looks like this one didn't get written.

---

*[SIL.Core/Xml/XmlUtils.cs, line 1209 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkGS--slmJkOevmFxZx:-KkLJdD_m3dnMkzD22ds:b-a1x2kv) ([raw file](https://github.com/sillsdev/libpalaso/blob/1f78b9a6eb46fb18887992ab750efdc0f92ff612/SIL.Core/Xml/XmlUtils.cs#L1209)):*
<details><summary><i>Previously, regnrand (Randy Regnier) wrote…</i></summary><blockquote>

Dunno. I just moved the code. I do know that FLEX has several more implementations of the IAttributeVisitor interface, and I'm reluctant to change so much behavior, without a better understanding of those return values.
</blockquote></details>

Oh, I think I get it now. We want to make sure we call visit on all attributes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-:-KkLJkrvHVs_ooKZU4td:b-gxrz16)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Reviewed 1 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/519)*
<!-- Sent from Reviewable.io -->

regnrand commented:
Review status: 6 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[SIL.Core.Tests/Xml/XmlUtilsTests.cs, line 368 at r2](https://reviewable.io:443/reviews/sillsdev/libpalaso/519#-KkLHuMumUw5qInW7-cj:-KkLa-PbPmO2uUKnUeS1:bv57foz) ([raw file](https://github.com/sillsdev/libpalaso/blob/6626cc904af08d674ef27f9322048efca49d50d1/SIL.Core.Tests/Xml/XmlUtilsTests.cs#L368)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Looks like this one didn't get written.
</blockquote></details>

Removed it. Its underlying method it was to test got removed. That method only did part of what the surviving method (XmlUtils.DecodeXmlAttribute) did, so it wasn't really needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/519)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/519)*
<!-- Sent from Reviewable.io -->
